kleros/action-callback-bots
# feat(bots): implement proof of humanity bot

# Description

This PR implements a new bot for the action callbacks of the [Proof Of Humanity smart contract](https://github.com/Proof-Of-Humanity/Proof-Of-Humanity/blob/master/contracts/ProofOfHumanity.sol).

This bot is a bit different from the others because it can leverage The Graph to more efficiently poll entities that match the contract's required conditions. I have structured the code into one module per action callback to clarify the relationship between the queries and the contract's required conditions.

E.g.

https://github.com/kleros/action-callback-bots/blob/68a27bc8a1f9400a68e155f999a728807966d6f0/src/bots/proof-of-humanity/change-state-to-pending.js#L3-L7

https://github.com/kleros/action-callback-bots/blob/68a27bc8a1f9400a68e155f999a728807966d6f0/src/bots/proof-of-humanity/change-state-to-pending.js#L33-L37

The PR also moves the repo to the new stable version of `web3`, because it caused compilation errors.

# How To Test This?

Run `npm start` and make sure the bot runs as expected in tandem with the others.

# Questions

I do not think these bots are resource-intensive enough to require us to run them on separate processes yet, but what is the throughput of the account sending the transactions?

Maybe it is time for each bot to get an individual account?
